internal/auth/oidc/repository_auth_method_update.go

```diff
@@ -17,32 +31,565 @@ import "context"
 // Options supported:
 //
 // * WithDryRun: when this option is provided, the auth method is retrieved from
-// the repo, updated based on the fieldMask, tested via Repository.TestAuthMethod
-// and any errors reported.  The updates are not peristed to the repository.
+// the repo, updated based on the fieldMask, tested via Repository.ValidateAuthMethod,
+// the results of the update are returned, and and any errors reported.  The
+// updates are not peristed to the repository.
 //
 // * WithForce: when this option is provided, the auth method is persistented in
-// the repository without testing it fo validity with Repository.TestAuthMethod.
+// the repository without testing it fo validity with Repository.ValidateAuthMethod.
 //
 // Successful updates must invalidate (delete) the Repository's cache of the
 // oidc.Provider for the AuthMethod.
-func (r *Repository) UpdateAuthMethod(ctx context.Context, m *AuthMethod, version uint32, fieldMaskPaths []string, _ ...Option) (*AuthMethod, error) {
+func (r *Repository) UpdateAuthMethod(ctx context.Context, am *AuthMethod, version uint32, fieldMaskPaths []string, opt ...Option) (*AuthMethod, int, error) {
 	const op = "oidc.(Repository).UpdateAuthMethod"
-	panic("to-do")
+	if am == nil {
+		return nil, db.NoRowsAffected, errors.New(errors.InvalidParameter, op, "missing auth method")
+	}
+	if am.AuthMethod == nil {
+		return nil, db.NoRowsAffected, errors.New(errors.InvalidParameter, op, "missing auth method store")
+	}
+	if am.PublicId == "" {
+		return nil, db.NoRowsAffected, errors.New(errors.InvalidParameter, op, "missing public id")
+	}
+
+	if err := validateFieldMask(fieldMaskPaths); err != nil {
+		return nil, db.NoRowsAffected, errors.Wrap(err, op)
+	}
+
+	dbMask, nullFields := dbcommon.BuildUpdatePaths(
+		map[string]interface{}{
+			"Name":             am.Name,
+			"Description":      am.Description,
+			"OperationalState": am.OperationalState,
```

Comment:
This should be a calculated field (see my comment above).

Reply:
I have concerns/questions which we should discuss.  

With that said, I think the UpdateAuthMethod(...) follows the spirt of the RFC as I read it.  

Unless an admin uses the WithForce option, the update operation determines if the auth method is valid, via AuthMethod.IsComplete and Repository.ValidateAuthMethod. IsComplete ensures the required configuration fields are populated and ValidateAuthMethod checks the configuration against the IdP's published discovery info.   

This implementation would appear to support the usability feature you've cited, by preventing an administrator from an embarrassing outage by updating auth method fields without ample warning that the auth method may be incomplete or misconfigured.   

Additionally, the implementation allows for one transaction/operation which can safely update an incomplete configuration and set the operational state to whatever the admin desires.  If admins prefer, they can split this workflow into separate operations: update then a 2nd transaction to change the operational state.    Admins can choose the workflow they wish (aka whatever supports their needs the best) vs a single workflow imposed by us.   Why not support both workflows, as long as both workflows provide administrators ample notification of possible configuration issues before inadvertent changes  are visible to users of the auth method?

Reply:
Sorry to disagree, but I don't think the UpdateAuthMethod(...) follows the spirit of the RFC, because one of the main goals I had when I wrote it was to prevent a provably misconfigured OIDC Auth Method from being exposed to end users.

What are the advantages of allowing end users to see and select an OIDC auth method that cannot work? What will be an end user's experience when 10 OIDC auth methods are listed but only 1 will work and they have no way of knowing which one? I understand the advantages of eliminating unnecessary steps in any workflow. However, I don't think this step is unnecessary since it prevents bad things from happening.

I'm happy to discuss any concerns you have. Let me know when you want to chat.

Reply:
Ty for the response and explanation!   I hope we can catch up in the morning before HEX kicks off.  I really appreciate the vision you have for protecting the end-user's experience.   

Unfortunately, I think the path we're headed down may make updates very difficult to reason about for the operator in some situations.  Imagine the scenario where I've created an auth method and "made it public", now I update the auth method using the WithForce option (because my provider's discovery info was and is never correct).  The update succeeds, but I'm blissfully unaware that my update had the side-affect of changing the auth method's operational state from public to inactive without raising an error.  

Perhaps we need some way to disambiguate an "incomplete" auth method from an auth method which doesn't pass the "Validate" tests we've prescribed.   This is the sort of ambiguous behavior which could be removed completely from the system, by simply making the update of OperationalStatus explicit.  You either update it directly, or you call one of the explicit domain functions to change the OperationalStatus (MakeInactive, MakePrivate, or MakePublic).  In my experience, system side-affects typically lead to system behavior which are difficult to reason about and properly disambiguate.  

As always, I'm looking forward to discussing this in real-time.  As always, our collaboration makes me a better engineer and I'm very grateful.

Reply:
see: https://github.com/hashicorp/boundary/pull/963/commits/2b110c144f910dee53f6d9e0b4bb091a11102b83